File: javascript/sublist/sublist.js
export class List {
  constructor(elements = []) {
    this.elements = elements;
  }

  compare(other) {
    const superlist = this.isSuperlistOf(other);
    const sublist = other.isSuperlistOf(this);
    if (superlist && sublist) {
      return 'EQUAL';
    } else if (superlist) {
      return 'SUPERLIST';
    } else if (sublist) {
      return 'SUBLIST';
    } else {
      return 'UNEQUAL';
    }
  }

  isSuperlistOf(other) {
    return this.elements.length >= other.elements.length &&
           (other.elements.length === 0 || [...this.elements.keys()].slice(0, this.elements.length - other.elements.length + 1)
                                                                    .some((i) => this.elements.slice(i, i + other.elements.length)
                                                                                              .every((e, idx) => e == other.elements[idx])));
  }
}

File: kotlin/sublist/src/main/kotlin/Sublist.kt
fun <T : Comparable<T>> List<T>.relationshipTo(other: List<T>): Relationship = when {
    size < other.size -> other.relationshipTo(this).let {
        if (it == Relationship.SUPERLIST) Relationship.SUBLIST else it
    }
    other.size == 0 -> when {
        size == 0 -> Relationship.EQUAL
        else -> Relationship.SUPERLIST
    }
    else -> when (windowed(other.size).any { it == other }) {
        true -> if (size == other.size) Relationship.EQUAL else Relationship.SUPERLIST
        false -> Relationship.UNEQUAL
    }
}
